plugin/storage/grpc/shared/streaming_writer.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2022 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shared
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.uber.org/atomic"
+
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/proto-gen/storage_v1"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+var (
+	_ spanstore.Writer = (*streamingSpanWriter)(nil)
+)
+
+const (
+	defaultMaxPoolSize = 1000
+)
+
+// streamingSpanWriter wraps storage_v1.StreamingSpanWriterPluginClient into spanstore.Writer
+type streamingSpanWriter struct {
+	client     storage_v1.StreamingSpanWriterPluginClient
+	streamPool chan storage_v1.StreamingSpanWriterPlugin_WriteSpanStreamClient
+	closed     *atomic.Bool
+}
+
+func newStreamingSpanWriter(client storage_v1.StreamingSpanWriterPluginClient) *streamingSpanWriter {
+	s := &streamingSpanWriter{
+		client:     client,
+		closed:     atomic.NewBool(false),
+		streamPool: make(chan storage_v1.StreamingSpanWriterPlugin_WriteSpanStreamClient, defaultMaxPoolSize),
+	}
+	return s
+}
+
+// WriteSpan write span into stream
+func (s *streamingSpanWriter) WriteSpan(ctx context.Context, span *model.Span) error {
+	stream, err := s.getStream(ctx)
+	if err != nil {
+		return fmt.Errorf("plugin getStream error: %w", err)
+	}
+	if err := stream.Send(&storage_v1.WriteSpanRequest{Span: span}); err != nil {
+		return fmt.Errorf("plugin Send error: %w", err)
+	}
+	s.putStream(stream)
```

Comment:
If I understand your code correctly, you pass a stream object through the channel over and over. That seems like a very roundabout way of synchronizing access to it (vs. just using a mutex). Also, it looks like it might open multiple streams (up to a 1000), e.g. when one is in use, the read from the channel will fall back to L87. Did you really want that?

Reply:
Yes, this is because stream.Send in gRPC is not concurrency safe (refer to: https://github.com/grpc/grpc-go/blob/master/stream.go#L123), but it seems that we have no way to bind streams and the upper layer goroutine worker, so a stream pool is implemented in this way.
Theoretically, the size of the stream pool will not exceed `collectorNumWorkers` because get and put are executed synchronously.
Btw, in my test, compared to unary writeSpan
1. Sharing a stream through mutex, although the CPU utilization will decrease, the throughput will decrease 
2. Sharing streams through the stream pool, the throughput will increase while the CPU utilization decreases.